### PR TITLE
improve registry client error handling

### PR DESCRIPTION
**Pull request by tomerd**
motivation: adjust client behavior to the spec

changes:
* try to parse server error as defined by sepc and produce a structured error with the information returned from the server
* add tests for error cases

### CODE REVIEW

**Comment by tomerd** [comment]
GH diff here is pretty horrible, but most of this is formatting / nesting changes. the actual change is handling unexpected error code more graceful, namely attempting to parse the server error and providing the information in a clean way instead of a generic "status code mismatch" error

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@yim-lee ptal

File: Sources/PackageRegistry/RegistryClient.swift [hunk 1]
```diff
@@ -120,28 +120,31 @@ public final class RegistryClient: Cancellable {
...
-                    try self.checkResponseStatusAndHeaders(
-                        response,
-                        expectedStatusCode: 200,
-                        expectedContentType: .json
-                    )
-
-                    guard let data = response.body else {
-                        throw RegistryError.invalidResponse
-                    }
+                    switch response.statusCode {
```
Comment (tomerd) [inline Sources/PackageRegistry/RegistryClient.swift:123]: this switch as the main change, it replaces the generic checkResponseStatusAndHeaders

File: Sources/PackageRegistry/RegistryClient.swift [hunk 2]
```diff
@@ -120,28 +120,31 @@ public final class RegistryClient: Cancellable {
...
-                        registry: registry,
-                        versions: versions,
-                        alternateLocations: alternateLocations?.map(\.url)
-                    )
+                        return PackageMetadata(
+                            registry: registry,
+                            versions: versions,
+                            alternateLocations: alternateLocations?.map(\.url)
+                        )
+                    default:
```
Comment (tomerd) [inline Sources/PackageRegistry/RegistryClient.swift:141]: this is where we try to parse the server error per the spec

**Comment by tomerd** [comment]
@swift-ci smoke test

File: Sources/PackageRegistry/RegistryClient.swift [hunk 3]
```diff
@@ -813,31 +868,29 @@ extension RegistryClient {
...
-        guard contentVersion == self.apiVersion.rawValue else {
-            throw RegistryError.invalidContentVersion(expected: self.apiVersion.rawValue, actual: contentVersion)
-        }
-
-        let contentType = response.headers.get("Content-Type").first
-        guard contentType?.hasPrefix(expectedContentType.rawValue) == true else {
-            throw RegistryError.invalidContentType(expected: expectedContentType.rawValue, actual: contentType)
-        }
-    }
+    /*
```
Comment (yim-lee) [inline Sources/PackageRegistry/RegistryClient.swift:878]: delete the code?

File: Sources/PackageRegistry/RegistryClient.swift [hunk 4]
```diff
@@ -896,6 +949,80 @@ extension RegistryClient {
...
+        guard let data = self.body else {
+            throw RegistryError.invalidResponse
+        }
+
+        return try decoder.decode(RegistryClient.ServerError.self, from: data)
+    }
+}
+
+extension HTTPClientResponse {
+    private func validateStatusCode(_ expectedStatusCodes: [Int]) throws {
```
Comment (yim-lee) [inline Sources/PackageRegistry/RegistryClient.swift:989]: Used?

**Review by yim-lee — APPROVED** [review]

**Pull request merged**